### DISCUSSION TITLE
Eleven Labs Custom Voice Text to Speech

### DISCUSSION THREAD

**Discussion by fsilverio-dev** [SemlerPDX/OpenAI-VoiceAttack-Plugin#3] [Ideas]
Hi,

This is a fantastic project you have going on here! I am learning to program and have always wanted to develop a personal assistant, the reason why I originally bought VoiceAttack many years ago.
I am anxious to try out this plugin myself and will give as much feedback as possible and help where I can!

I have been experimenting with Auto-GPT lately
 and they use a simple API call to have a custom AI-generated voice,
Could this potentially be Implemented in the Plugin?

[speak.zip](https://github.com/SemlerPDX/OpenAI-VoiceAttack-Plugin/files/11594380/speak.zip)

In the attachment the script I'm referring to, I have little knowledge of how to implement this, any guidance would be appreciated!

**Comment by SemlerPDX** [comment]
That is a great question, and the answer is YES ...with a "but". Such a system should not be included natively, as that is not the only voice generation API available, and is not a part of the OpenAI API which is the purpose of this plugin. That being said, I have built in options for using whatever voice is desired to speak the returned responses.

The proper way to implement this with use of this plugin would be inside the required "((OpenAI Speech))" command in VoiceAttack, which is where all speech produced by the OpenAI Plugin for VoiceAttack occurs.  This achieves a separation of concerns for the plugin itself and the actual voice (and options for that voice), up to and including use of other systems such as that elevenlabs.io's API python script achieves, or perhaps Microsoft's Cognitive Speech through the "**Microsoft Cognitive Speech plug-in for Voice Attack**" which can be found here:
[https://github.com/jamescl604/MSCognitiveSpeechForVoiceAttack](https://github.com/jamescl604/MSCognitiveSpeechForVoiceAttack)

Some users may have voices already installed, and so for them, it is as simple as the plugin doing what it already does, use a text-to-speech action in a single command in the profile using this plugin. For others, they can simply disable that action and introduce their own system which uses some means to speak the text contained in the [`OpenAI_TTS_Response`](https://github.com/SemlerPDX/OpenAI-VoiceAttack-Plugin/wiki/All-VoiceAttack-Variables#openai_tts_response) text variable.
